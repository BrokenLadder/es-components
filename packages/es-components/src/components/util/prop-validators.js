import isParseableDate from './isParseableDate';

export function isPropValidDate(propsList, propName, componentName) {
  const date = propsList[propName];
  if (date) {
    if (!isParseableDate(date)) {
      return new Error(`The pre-selected date is not valid. Please enter a parseable date.
        See http://en.wikipedia.org/wiki/ISO_8601 for valid date formats.
      `);
    }
    return null;
  }
  return null;
}
