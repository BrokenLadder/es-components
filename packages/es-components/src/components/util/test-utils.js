import React from 'react';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from 'react-testing-library';

export function renderWithTheme(component) {
  return render(
    <ThemeProvider theme={viaTheme}>
      {React.cloneElement(component)}
    </ThemeProvider>
  )
}