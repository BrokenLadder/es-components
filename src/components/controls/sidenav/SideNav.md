```
<div style={{display: 'flex'}}>
  <div style={{width: '30%', marginRight: '4em'}}>
    <h3>Standard</h3>
    <SideNav defaultSelected="home">
      <SideNav.Item id="home"><a>Home</a></SideNav.Item>
      <SideNav.Item id="cart">Cart</SideNav.Item>
      <SideNav.Item id="help">Help</SideNav.Item>
      <SideNav.Item id="info"><a>About</a></SideNav.Item>
      <SideNav.Item id="info" isDisabled>Disabled</SideNav.Item>
    </SideNav>
  </div>

  <div style={{width: '30%'}}>
    <h3>Alternate Style</h3>
    <SideNav defaultSelected="home" altStyle>
      <SideNav.Item id="home"><a>Home</a></SideNav.Item>
      <SideNav.Item id="cart">Cart</SideNav.Item>
      <SideNav.Item id="help">Help</SideNav.Item>
      <SideNav.Item id="info"><a>About</a></SideNav.Item>
      <SideNav.Item id="info" isDisabled>Disabled</SideNav.Item>
    </SideNav>
  </div>
</div>
```

Setting ```onItemSelected``` will fire the function when a ```SideNav.Item``` is clicked.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" onItemSelection={ (id) => {alert(id); }}>
    <SideNav.Item id="home">Home</SideNav.Item>
    <SideNav.Item id="cart">Cart</SideNav.Item>
    <SideNav.Item id="info" isDisabled>Disabled</SideNav.Item>
  </SideNav>
</div>
```

Set ```onNavClick``` on an individual ```SideNav.Item```.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" altStyle>
    <SideNav.Item id="home" onNavClick={ () => (alert('Time to go Home!'))}>Home</SideNav.Item>
    <SideNav.Item id="cart" onNavClick={ () => (alert('You clicked the Cart!'))}>Cart</SideNav.Item>
    <SideNav.Item id="info" isDisabled>Disabled</SideNav.Item>
  </SideNav>
</div>
```

Manage selected item state manually using ```selected``` instead of ```defaultSelected```.
```
<div style={{width: '30%'}}>
  <SideNav selected={state.selected} onItemSelection={ (id) => {setState({selected: id})} }>
    <SideNav.Item id="home"><a>Home</a></SideNav.Item>
    <SideNav.Item id="cart">Cart</SideNav.Item>
    <SideNav.Item id="info"><a>Help</a></SideNav.Item>
  </SideNav>
</div>
```
