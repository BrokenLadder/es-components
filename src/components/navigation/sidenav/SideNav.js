import React, { Children, Component, cloneElement } from 'react';
import PropTypes from 'prop-types';
import { colors } from '../../theme';
import styled from 'styled-components';
import { noop } from 'lodash';

import NavItem from './NavItem';

const NavStyled = styled.nav`
  background-color: ${colors.white};
  box-shadow: 2px 2px 5px ${colors.grayDark};

  ul {
    list-style: none;
    margin: 0;
    padding: 0;
  }
`;

class SideNav extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: props.defaultSelected,
      defaultSelected: props.defaultSelected
    };
  }

  onNavClick = (navId = null) => {
    const { onItemSelected = noop } = this.props;

    if (this.state.defaultSelected) {
      this.setState({ selected: navId }, () => {
        onItemSelected(navId);
      });
    } else {
      onItemSelected(navId);
    }
  };

  render() {
    const { altStyle, children, selected } = this.props;
    return (
      <NavStyled>
        <ul>
          {Children.toArray(children).map(child => {
            if (child !== null && child.type === NavItem) {
              const currentSelected = this.state.defaultSelected
                ? this.state.selected
                : selected;
              return cloneElement(child, {
                altStyle,
                highlightedId: currentSelected,
                onClick: this.onNavClick
              });
            }
            return child;
          })}
        </ul>
      </NavStyled>
    );
  }
}

SideNav.propTypes = {
  /** Use the alternate nav style */
  altStyle: PropTypes.bool,
  children: PropTypes.node,
  /** Use to manually select nav item by id, stateless mode */
  selected: PropTypes.string,
  /** Use to set a default nav, stateful mode */
  defaultSelected: PropTypes.string,
  /** Function called when a nav item is clicked */
  onItemSelected: PropTypes.func
};

SideNav.Item = NavItem;

export default SideNav;
