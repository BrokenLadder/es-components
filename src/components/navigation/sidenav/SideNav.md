```
<div style={{display: 'flex'}}>
  <div style={{width: '30%', marginRight: '4em'}}>
    <h3>Standard</h3>
    <SideNav defaultSelected="home">
      <SideNav.Item id="home" title="Home" targetUrl="#home" />
      <SideNav.Item id="cart" title="Cart">Cart</SideNav.Item>
      <SideNav.Item id="help" title="Help" />
      <SideNav.Item id="info" title="About" />
      <SideNav.Item id="disabled" title="Diabled" isDisabled />
    </SideNav>
  </div>

  <div style={{width: '30%'}}>
    <h3>Alternate Style</h3>
    <SideNav defaultSelected="home" altStyle>
      <SideNav.Item id="home" title="Home" />
      <SideNav.Item id="cart" title="Cart">Cart</SideNav.Item>
      <SideNav.Item id="help" title="Help" />
      <SideNav.Item id="info" title="About" />
      <SideNav.Item id="disabled" title="Diabled" isDisabled />
    </SideNav>
  </div>
</div>
```

Assign a ```targetUrl``` to set the nav link href location. Use ```isExternalLink``` to open the link in a new browser window.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home">
    <SideNav.Item id="home" targetUrl="/" title="Home" onNavClick={ () => (alert('Time to go Home!'))} />
    <SideNav.Item id="medicare" targetUrl="https://medicare.oneexchange.com" title="Medicare" isExternalLink />
    <SideNav.Item id="alerts" targetUrl="#alert" title="Alert Section" />
  </SideNav>
</div>
```

Setting ```onItemSelected``` will execute the function when a ```SideNav.Item``` is clicked.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" onItemSelected={ (id) => {alert(id); }}>
    <SideNav.Item id="home" title="Home" />
    <SideNav.Item id="cart" title="Cart" />
    <SideNav.Item id="info" isDisabled title="Disabled" />
  </SideNav>
</div>
```

Set ```onNavClick``` on an individual ```SideNav.Item```.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" altStyle>
    <SideNav.Item id="home" onNavClick={ () => (alert('Time to go Home!'))} title="Home" />
    <SideNav.Item id="cart" onNavClick={ () => (alert('You clicked the Cart!'))} title="Cart" />
    <SideNav.Item id="info" isDisabled title="Disabled" />
  </SideNav>
</div>
```

Manage selected item state manually using ```selected``` instead of ```defaultSelected```.
```
<div style={{width: '30%'}}>
  <SideNav selected={state.selected} onItemSelected={ (id) => {setState({selected: id})} }>
    <SideNav.Item id="home" title="Home" />
    <SideNav.Item id="cart" title="Cart" />
    <SideNav.Item id="help" title="Help" />
  </SideNav>
</div>
```
