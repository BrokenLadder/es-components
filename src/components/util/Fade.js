import React from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-overlays/lib/Transition';
import { injectGlobal } from 'styled-components';

/* eslint-disable no-unused-expressions */
injectGlobal`
  .default-transition-out {
    opacity: 0;
    transition: opacity 300ms linear;
  }

  .default-transition-in {
    opacity: 1;
  }
`;
/* eslint-enable no-unused-expressions */

const Fade = props => {
  const {
    in: transitionIn,
    timeout,
    transitionClassIn,
    transitionClassOut,
    ...otherProps
  } = props;

  return (
    <Transition
      className={transitionClassOut}
      enteredClassName={transitionClassIn}
      enteringClassName={transitionClassIn}
      in={transitionIn}
      unmountOnExit
      timeout={timeout}
      transitionAppear
      {...otherProps}
    />
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  in: PropTypes.bool,
  transitionClassIn: PropTypes.string,
  transitionClassOut: PropTypes.string,
  timeout: PropTypes.number
};

Fade.defaultProps = {
  transitionClassIn: 'default-transition-in',
  transitionClassOut: 'default-transition-out',
  timeout: 5000
};

export default Fade;
