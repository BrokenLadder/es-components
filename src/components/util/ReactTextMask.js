import React from 'react';
import ReactTextMask from 'react-text-mask';

// the component was using onInput which made testing more difficult.
// switched to onChange and put in a pull request so hopefully we can
// get rid of this soon.
ReactTextMask.prototype.render = function render() {
  const props = { ...this.props };

  delete props.mask;
  delete props.guide;
  delete props.pipe;
  delete props.placeholderChar;
  delete props.keepCharPositions;
  delete props.onChange;
  delete props.value;
  delete props.showMask;

  return (
    <input
      {...props}
      onChange={event => this.onChange(event)}
      defaultValue={this.props.value}
      ref={inputElement => (this.inputElement = inputElement)}
    />
  );
};

export default ReactTextMask;
