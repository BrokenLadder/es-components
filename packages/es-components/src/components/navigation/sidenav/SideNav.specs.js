/* eslint-env jest */

import React from 'react';
import { renderWithTheme } from '../../util/test-utils';

import SideNav from './SideNav';

let instanceToRender;
const onItemSelected = jest.fn();
const onClick = jest.fn();

beforeEach(() => {
  onItemSelected.mockClear();
  onClick.mockClear();

  instanceToRender = (
    <SideNav onItemSelected={navId => onItemSelected(navId)}>
      <SideNav.Item
        id="home"
        className="home"
        onClick={navId => onClick(navId)}
        targetUrl="/home"
      >
        Home
      </SideNav.Item>
      <SideNav.Item id="cart" className="cart">
        Cart
      </SideNav.Item>
      <SideNav.Item
        id="external"
        className="external"
        targetUrl="http://www.google.com"
        isExternalLink
      >
        External Link
      </SideNav.Item>
      <SideNav.Item
        id="info"
        className="info"
        onClick={() => onClick()}
        isDisabled
      >
        Disabled
      </SideNav.Item>
    </SideNav>
  );
});

it('renders as expected', () => {
  const { container } = renderWithTheme(instanceToRender);
  expect(container).toMatchSnapshot();
});

it('executes onItemSelected with the id of the nav item clicked', () => {
  const { getByText } = renderWithTheme(instanceToRender);
  getByText('Cart').click();
  expect(onItemSelected).toBeCalledWith('cart');
});

it('executes onClick when nav item clicked', () => {
  const { getByText } = renderWithTheme(instanceToRender);
  getByText('Home').click();
  expect(onClick).toBeCalledWith('home');
});

it('disabled item prevents onclick functions', () => {
  const { getByText } = renderWithTheme(instanceToRender);
  getByText('Disabled').click();
  expect(onClick).not.toBeCalled();
  expect(onItemSelected).not.toBeCalled();
});
