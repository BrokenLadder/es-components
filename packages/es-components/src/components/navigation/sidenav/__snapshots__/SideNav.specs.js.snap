// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawer renders as expected 1`] = `
<nav
  className="sc-ifAKCX kXLRTf"
>
  <ul>
    <li>
      <a
        className="home sc-bwzfXH bFUGaW"
        href="/home"
        onClick={[Function]}
      >
        <span
          className="sc-bxivhb khxFsg"
        >
          Home
        </span>
        <i
          aria-hidden={true}
          className="sc-bdVaJa cMtiSf"
          content="61f"
          fontFamily="indv-mkt-icons"
          fontSize="inherit"
        />
      </a>
    </li>
    <li>
      <a
        className="cart sc-bwzfXH bFUGaW"
        href="#/"
        onClick={[Function]}
      >
        <span
          className="sc-bxivhb khxFsg"
        >
          Cart
        </span>
        <i
          aria-hidden={true}
          className="sc-bdVaJa cMtiSf"
          content="61f"
          fontFamily="indv-mkt-icons"
          fontSize="inherit"
        />
      </a>
    </li>
    <li>
      <a
        className="external sc-bwzfXH bFUGaW"
        href="http://www.google.com"
        onClick={[Function]}
        target="_blank"
      >
        <span
          className="sc-bxivhb khxFsg"
        >
          External Link
        </span>
        <i
          aria-hidden={true}
          className="sc-bdVaJa cMtiSf"
          content="61f"
          fontFamily="indv-mkt-icons"
          fontSize="inherit"
        />
      </a>
    </li>
    <li>
      <a
        className="info sc-bwzfXH evWADI"
        href="#/"
        onClick={[Function]}
      >
        <span
          className="sc-bxivhb khxFsg"
        >
          Disabled
        </span>
        <i
          aria-hidden={true}
          className="sc-bdVaJa cMtiSf"
          content="61f"
          fontFamily="indv-mkt-icons"
          fontSize="inherit"
        />
      </a>
    </li>
  </ul>
</nav>
`;
